File: lib/to_mdx/parsers/open_document/style/style_node.rb
# frozen_string_literal: true

require_relative "../base"

module ToMdx
  module Parsers
    module Style
      class StyleNode < Base
        attr_reader :style_name, :parent_style_name, :display_name

        def initialize(element = nil, context = {})
          super(element, context)
          @style_name = element&.attributes&.[]("style:name")
          @parent_style_name = element&.attributes&.[]("style:parent-style-name")
          @display_name = element&.attributes&.[]("style:display-name")
        end

        def to_css
          return {} unless @style_name
          { @style_name => children.map(&:to_css).reduce({}, :merge) }
        end

        def with_context(context)
          context.merge(
            style_family: element&.attributes&.[]("style:family"),
          )
        end

        register_for "style:style"
        register_for "style:default-style"
      end
    end
  end
end

File: src/parsers/open-document/style/style_node.ts
import type { ParseContext } from "../../base_element.ts";
import { BaseElement } from "../../base_element.ts";

export class StyleNode extends BaseElement {
  private name: string | null;

  constructor(element: Element, context: ParseContext, parent: BaseElement | null) {
    super(element, context, parent);
    this.name = this.attr("style:name");
  }

  toMdx(): string {
    return "";
  }

  toString(): string {
    return "";
  }

  toCss(): Record<string, Record<string, string>> {
    if (!this.name) {
      return {};
    }

    const childStyles = this.children
      .map((child) => {
        if ("toCss" in child && typeof child.toCss === "function") {
          return (child as unknown as { toCss: () => Record<string, string> }).toCss();
        }
        return {};
      })
      .reduce((acc, styles) => ({ ...acc, ...styles }), {});

    return { [this.name]: childStyles };
  }

  protected withContext(context: ParseContext): ParseContext {
    return {
      ...context,
      style_family: this.attr("style:family"),
    };
  }

  static {
    BaseElement.registerFor.call(this, "style:style", "style:default-style");
  }
}
